### PR TITLE
[v9] Reorganize `FirebaseCore*` modules

### PR DESCRIPTION
**Pull request by ncooke3**
```mermaid
graph BT
    G[/Heartbeat Logging Source\] --- A
    F[/Future parts of Core written in Swift\] -.- A
    A{{FirebaseCoreInternal}} --> B(FirebaseCore)
    B --> C{{FirebaseCoreExtension}}
    B --> D((Clients))
    C --> E(Firebase Product SDKs)
    B --> E
    E --> D
```
**Key**
- Arrows indicate dependencies 
- Lines indicate the code is _contained_ by the module above it
- Trapezoid nodes represent the bottom of the dep chain within Firebase
- Hexagon nodes are for modules designed for internal Firebase use
- Rounded rectangle nodes are for modules with API that's publicly exposed to customers

#no-changelog

### CODE REVIEW

**Comment by ncooke3** [comment]
The `heartbeat_logging` workflows are failing because they are out of date. 

Not super sure what's happening in https://github.com/firebase/firebase-ios-sdk/runs/5619184906?check_suite_focus=true...

File: Crashlytics/Crashlytics/DataCollection/FIRCLSDataCollectionArbiter.m [hunk 1]
```diff
@@ -20,7 +20,7 @@
 #import "FBLPromises.h"
 #endif
 
-#import "FirebaseCore/Internal/FirebaseCoreInternal.h"
+#import "FirebaseCore/Extension/FirebaseCoreInternal.h"
```
Comment (paulb777) [inline Crashlytics/Crashlytics/DataCollection/FIRCLSDataCollectionArbiter.m:23]: Now that module imports are ok, we might want to delete FirebaseCoreInternal.h in favor of `@import FirebaseCoreExtension;`  - although I'm not sure we have a g3 strategy for that and may need to keep FirebaseCoreInternal.h at least on the g3 side?

File: scripts/spm_test_schemes/FirebaseCoreInternalTests.xcscheme [hunk 2]
```diff
@@ -1,26 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <Scheme
-   LastUpgradeVersion = "1300"
+   LastUpgradeVersion = "1320"
    version = "1.3">
    <BuildAction
       parallelizeBuildables = "YES"
       buildImplicitDependencies = "YES">
-      <BuildActionEntries>
```
Comment (paulb777) [inline scripts/spm_test_schemes/FirebaseCoreInternalTests.xcscheme:8]: Keep it buildable

File: scripts/pod_lib_lint.rb [hunk 3]
```diff
@@ -68,6 +68,7 @@ def main(args)
   podspec_file = pod_args[0]
   # Figure out which dependencies are local
   deps = find_local_deps(podspec_file, ignore_local_podspecs.to_set)
+  puts 'podspec_file: ' + podspec_file + '\n podspec_file_deps: ' + deps.join(', ')
```
Comment (paulb777) [inline scripts/pod_lib_lint.rb:71]: Don't commit these here

Comment (ncooke3) [reply scripts/pod_lib_lint.rb]: Reverted in [b704b09](https://github.com/firebase/firebase-ios-sdk/pull/9482/commits/b704b098b6c556806b8d6366265fba774105c885)

File: scripts/spm_test_schemes/FirebaseCoreInternalTests.xcscheme (hunk 2 above)
Comment (ncooke3) [reply scripts/spm_test_schemes/FirebaseCoreInternalTests.xcscheme]: Good catch! Done in [bb32a8b](https://github.com/firebase/firebase-ios-sdk/pull/9482/commits/bb32a8b550fcf3810295691b82941392d3278e40)

File: .github/workflows/core_internal.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        flags: [
+          '--use-static-frameworks'
+        ]
+    needs: pod-lib-lint
+    steps:
+    - uses: actions/checkout@v2
+    - name: Setup Bundler
+      run: scripts/setup_bundler.sh
+    - name: PodLibLint CoreInternal Cron
+      run: scripts/third_party/travis/retry.sh scripts/pod_lib_lint.rb FirebaseCoreInternal.podspec --platforms=${{ matrix.target }} ${{ matrix.flags }}
```
Comment (paulb777) [inline .github/workflows/core_internal.yml:79]: Add newline

File: Crashlytics/Crashlytics/DataCollection/FIRCLSDataCollectionArbiter.m (hunk 1 above)
Comment (paulb777) [reply Crashlytics/Crashlytics/DataCollection/FIRCLSDataCollectionArbiter.m:23]: I'm ok to defer this decision.

File: Package.swift [hunk 5]
```diff
@@ -149,7 +149,7 @@ let package = Package(
       url: "https://github.com/google/GoogleAppMeasurement.git",
       // Note that CI changes the version to the head of main for CI.
       // See scripts/setup_spm_tests.sh.
-      .exact("8.12.0")
+      .branch("main")
```
Comment (paulb777) [inline Package.swift:152]: Shouldn't be checked in. We really need to do https://github.com/firebase/firebase-ios-sdk/issues/9072

File: .github/workflows/core_internal.yml [hunk 6]
```diff
@@ -0,0 +1,79 @@
+name: core_internal
```
Comment (paulb777) [inline .github/workflows/core_internal.yml:1]: Don't we need a `core_extension.yml` also?

**Review by paulb777 — COMMENTED** [review]
ReleaseTooling changes are also missing, but you may want to wait for #9471 to merge

File: Package.swift (hunk 5 above)
Comment (ncooke3) [reply Package.swift]: Thanks. Done in 0826e4b42c1d929a4b926180287d428ecb8f7a58.

File: .github/workflows/core_internal.yml (hunk 4 above)
Comment (ncooke3) [reply .github/workflows/core_internal.yml]: Done in 0826e4b42c1d929a4b926180287d428ecb8f7a58.

File: .github/workflows/core_internal.yml (hunk 6 above)
Comment (ncooke3) [reply .github/workflows/core_internal.yml:1]: Yes, added in 0826e4b42c1d929a4b926180287d428ecb8f7a58.

Currently, the workflow is set to lint the `FirebaseCoreExtension.podspec` (which currently only contains source headers). IIUC, I can't just build the source for SPM (no scheme generates for the `FirebaseCoreExtension` target). I could think of three options:
1. No op– don't run `spm` workflows for FirebaseCoreExtension.
2. Add a `FirebaseCoreExtensionTesting` scheme that has no tests but builds `FirebaseCoreExtension`.
3. Add a `FirebaseCoreExtensionTesting` scheme and rearrange the `FirebaseCore/Extension` directory and add API build tests for both Swift and ObjC? The package modifications would look something like:
image

WDYT?

Comment (paulb777) [reply .github/workflows/core_internal.yml:1]: Since it's only headers, `pod lib lint` testing should be sufficient.

Comment (ncooke3) [reply .github/workflows/core_internal.yml:1]: Done in [e9483fa](https://github.com/firebase/firebase-ios-sdk/pull/9482/commits/e9483fa2565916206cf61e0763ac52f9354e5db1)

**Review by paulb777 — APPROVED** [review]

**Pull request merged**